File: demo/max_subsequence_sum.c
//
// Created by James Wang on 2021/7/6.
//
#include <stdio.h>

// 简单穷举
__attribute__((unused)) int max_subsequence_sum1(const int arr[], int n)
{
	int this_sum, max_sum = 0;
	for (int i = 0; i < n; i++)
		for (int j = i; j < n; j++) {
			this_sum = 0;
			for (int k = i; k <= j; k++)
				this_sum += arr[k];
			if (this_sum > max_sum)
				max_sum = this_sum;
		}

	return max_sum;
}

// 在构造子序列的同时求和
__attribute__((unused)) int max_subsequence_sum2(const int *arr, int n)
{
	int this_sum, max_sum = 0;
	for (int i = 0; i < n; i++) {
		this_sum = 0;
		for (int j = i; j < n; j++) {
			this_sum += arr[j];
			if (this_sum > max_sum)
				max_sum = this_sum;
		}
	}

	return max_sum;
}

static int max3(const int left, const int right, const int center)
{
	int max;
	if (left > right)
		max = left;
	else
		max = right;
	if (center > max)
		max = center;

	return max;
}

// 采用分而治之，递归的思想
static int max_sub_sum(const int arr[], int left, int right)
{
	int max_left_sum, max_right_sum;
	int max_left_border_sum, max_right_border_sum;
	int left_border_sum, right_border_sum;
	int center, i;

	if (left == right) // 基准情形
		if (arr[left] > 0)
			return arr[left];
		else
			return 0;

	center = (left + right) / 2;
	max_left_sum = max_sub_sum(arr, left, center);
	max_right_sum = max_sub_sum(arr, center + 1, right);

	max_left_border_sum = 0;
	left_border_sum = 0;
	for (i = center; i >= left; i--) {
		left_border_sum += arr[i];
		if (left_border_sum > max_left_border_sum)
			max_left_border_sum = left_border_sum;
	}

	max_right_border_sum = 0;
	right_border_sum = 0;
	for (i = center + 1; i <= right; i++) {
		right_border_sum += arr[i];
		if (right_border_sum > max_right_border_sum)
			max_right_border_sum = right_border_sum;
	}

	return max3(max_left_sum, max_right_sum,
		    max_left_border_sum + max_right_border_sum);
}

__attribute__((unused)) int max_subsequence_sum3(const int arr[], int n)
{
	return max_sub_sum(arr, 0, n - 1);
}

// 几乎完美的算法：仅需要常量空间并以线性时间运行的联机算法
int max_subsequence_sum4(const int arr[], int n)
{
	int this_sum = 0, max_sum = 0;
	for (int i = 0; i <= n; i++) {
		this_sum += arr[i];
		if (this_sum > max_sum)
			max_sum = this_sum;
		else if (this_sum <
			 0) // 0很关键，因为我们认为子序列的和一定是大于0的
			this_sum =
				0; // 所以当子序列和小于0时，重置其为0,抛弃索引i之前的所有元素
	}

	return max_sum;
}

int main(void)
{
	//    int array[] = {-2, 11, -4, 13, -5, -2};
	//    int array[] = {4, -3, 5, -2, -1, 2, 6, -2};
	int array[] = { 9, 9, -9, -9 };
	int size = sizeof array / sizeof *array;
	//    int max_sum = max_subsequence_sum1(array, size);
	//    int max_sum = max_subsequence_sum2(array, size);
	//    int max_sum = bigO_logn(array, 0, size - 1);
	//    int max_sum = max_subsequence_sum3(array, size);
	int max_sum = max_subsequence_sum4(array, size);
	printf("max subsequence sum: %d\n", max_sum);
}

File: Python/demo/max_subsequence_sum.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

def max_subsequence_sum1(arr):
    """
    简单穷举
    """
    max_sum = 0
    for i in range(0, len(arr)):  # 遍历所有元素作为子序列的第一项
        for j in range(i, len(arr)):  # 确定子序列的最后一项
            this_sum = 0
            for k in arr[i: j + 1]:  # 遍历子序列求和
                this_sum += k
            if this_sum > max_sum:
                max_sum = this_sum

    return max_sum


def max_subsequence_sum2(arr):
    """
    在构造子序列的同时求和
    """
    max_sum = 0
    for i in range(0, len(arr)):  # 遍历所有元素作为子序列的第一项
        this_sum = 0
        for j in range(i, len(arr)):  # 确定子序列最后一项
            this_sum += arr[j]  # 求和
            if this_sum > max_sum:
                max_sum = this_sum

    return max_sum


def max3(left, right, center):
    return max((left, right, center))


def max_sub_sum(arr, left, right):
    """
    采用分而治之，递归的思想
    """
    # 基准情形
    if left == right:
        return arr[left] if arr[left] > 0 else 0

    center = (left + right) // 2  # Python 的整除符号为 //
    max_left_sum = max_sub_sum(arr, left, center)
    max_right_sum = max_sub_sum(arr, center + 1, right)

    left_border_sum = max_left_border_sum = 0
    for i in reversed(arr[left: center + 1]):
        left_border_sum += i
        if left_border_sum > max_left_border_sum:
            max_left_border_sum = left_border_sum

    right_border_sum = max_right_border_sum = 0
    for i in arr[center + 1: right + 1]:
        right_border_sum += i
        if right_border_sum > max_right_border_sum:
            max_right_border_sum = right_border_sum

    return max3(max_left_sum, max_right_sum, max_left_border_sum + max_right_border_sum)


def max_subsequence_sum3(arr):
    return max_sub_sum(arr, 0, len(arr) - 1)


def max_subsequence_sum4(arr):
    """
    几乎完美的算法：仅需要常量空间并以线性时间运行的联机算法
    """
    max_sum = this_sum = 0
    for i in arr:
        this_sum += i
        if this_sum > max_sum:
            max_sum = this_sum
        elif this_sum < 0:
            this_sum = 0

    return max_sum


def test():
    array = [-2, 11, -4, 13, -5, -2]
    # array = [9, 9, -9, -9]
    # max_sum = max_subsequence_sum1(array)
    # max_sum = max_subsequence_sum2(array)
    # max_sum = max_subsequence_sum3(array)
    max_sum = max_subsequence_sum4(array)
    print(f'max subsequence sum: {max_sum}')


if __name__ == '__main__':
    test()